### PR TITLE
Utilities for multivariate gaussians

### PR DESCRIPTION
**Pull request by richfitz**
Some utilities for working with multivariate gaussians:

* compute the density
* compute the derivative of the density with respect to the parameters (assuming zero mean)

Currently merging into #1 because it contains those commits due to the support for sampling from a multivariate gaussian

### CODE REVIEW

File: R/distributions.R [hunk 1]
```diff
@@ -13,3 +13,31 @@ make_rmvnorm <- function(vcv) {
     x + drop(rnorm(n) %*% r)
   }
 }
+
+
+ldmvnorm <- function(x, vcv) {
```
Comment (M-Kusumgar) [inline R/distributions.R:18]:
```suggestion
# log density multivariate normal
ldmvnorm <- function(x, vcv) {
```

File: R/distributions.R [hunk 2]
```diff
@@ -13,3 +13,31 @@ make_rmvnorm <- function(vcv) {
...
+  dec <- base::chol(vcv)
+  constant <- -sum(log(diag(dec))) - 0.5 * nrow(vcv) * log(2 * pi)
+  function(x) {
+    constant - 0.5 * sum(backsolve(dec, x, transpose = TRUE)^2)
+  }
+}
+
+
+## Not the world's most beautiful name.
+dldmvnormdx <- function(x, vcv) {
```
Comment (M-Kusumgar) [inline R/distributions.R:33]: perhaps?
```suggestion
deriv_ldmvnorm <- function(x, vcv) {
```

File: R/distributions.R [hunk 3]
```diff
@@ -13,3 +13,31 @@ make_rmvnorm <- function(vcv) {
...
+  make_dldmvnormdx(vcv)(x)
+}
+
+
+make_dldmvnormdx <- function(vcv) {
```
Comment (M-Kusumgar) [inline R/distributions.R:38]: similarly
```suggestion
make_deriv_ldmvnorm <- function(vcv) {
```

File: tests/testthat/test-distributions.R [hunk 4]
```diff
@@ -29,3 +29,27 @@ test_that("rvnorm produces correct expectation", {
...
   expect_equal(var(r), vcv, tolerance = 0.01)
 })
+
+
+test_that("can compute the log density of a multivariate normal", {
+  vcv <- matrix(c(4, 2, 2, 3), ncol = 2)
+  set.seed(1)
+  x <- matrix(rnorm(20, sd = 10), ncol = 2)
+  expect_equal(apply(x, 1, make_ldmvnorm(vcv)),
+               mvtnorm::dmvnorm(x, sigma = vcv, log = TRUE))
```
Comment (M-Kusumgar) [inline tests/testthat/test-distributions.R:39]: is our implementation of the `rmvnorm` better? wondering why we implemented that in the distributions file

**Review by M-Kusumgar — CHANGES_REQUESTED** [review]
Looks good, would be nice if we had a little bit more explanation on the variable names!

File: tests/testthat/test-distributions.R (hunk 4 above)
Comment (richfitz) [reply tests/testthat/test-distributions.R:39]: Ours caches the decomposition. Later ours will also be translated into C++

**Review by M-Kusumgar — APPROVED** [review]
LGTM :)

**Pull request merged**